### DISCUSSION TITLE
Event.On cannot use expiration logic?

### DISCUSSION THREAD

**Discussion by GeekStocks** [wailsapp/wails#598] [Q&A]
I have a need to respond to an event one (and **only** one) time. In other frameworks I am used to seeing `.On` and `.Off` logic so that a listener can be removed after the first run (in my use case). I was surprised to see there is no `.Off` in Wails and I was going to propose a pull request to add it. _That is, until I read the rather fancy expiration logic already written in_ `Manager.Start()`:

```
// Update listen counter
if listener.counter > 0 {
	listener.counter = listener.counter - 1
	if listener.counter == 0 {
		listener.expired = true
	}
}
```

So since you've already implemented a way to expire a listener after x runs, why does `.On` not expose it here, but rather just hard codes a zero?

```
// at lib/event/manager.go

// On adds a listener for the given event
func (e *Manager) On(eventName string, callback func(...interface{})) {
	// Add a persistent eventListener (counter = 0)
	e.addEventListener(eventName, callback, 0)
}
```

Have I missed the proper way to `.On` an event x number of times? **Thanks!**

**Comment by leaanthony** [comment]
I'd created `Once()` for that in the JS runtime. Looks like it was never ported to the Go runtime. Good catch! Let's fix that.

**Reply by leaanthony** [reply]
@GeekStocks This should be fixed now. Please try `wails update -pre` for the latest pre-release

**Reply by GeekStocks** [reply]
`-pre2` updated and compiled fine. Sad to say my tests failed however. `Once` does **not** expire the handler after one run as expected.

This is a simple "React tells Go when App is loaded, Go responds with the next thing to do" event sequence like so:

```
// startup.go
func (s *StartUp) WailsInit(runtime *wails.Runtime) error {
	s.runtime = runtime

	// listen ONE TIME for the JS client (React "App" component) to Emit a "loaded" event
	runtime.Events.Once("reactAppLoaded_", s.handleReactAppLoaded)
	return nil
}
```
and the handler...
```
func (s *StartUp) handleReactAppLoaded(data ...interface{}) {
	s.runtime.Events.Emit("_loadFile", s.loadFile) // notify frontend to request a file...open

	// DEBUG TEST FOR THE NEW .Once METHOD
	time.AfterFunc(time.Second*5, func() {
		s.runtime.Events.Emit("reactAppLoaded_") // THIS SHOULD NOT SUCCEED, BUT IT DOES
		fmt.Println("DEBUGGING EMIT SENT")
	})
}
```
My frontend `console.logs()` every time it hears `"_loadFile"` emitted, and every 5 seconds it is printing. It should print just once and then it is expected that subsequent Emits of `"_loadFile"` would silently fail as no handler would be on the stack (or in Wails case, be unexpired).

**Reply by GeekStocks** [reply]
> @GeekStocks This should be fixed now. Please try `wails update -pre` for the latest pre-release

Hey @leaanthony shouldn't `.Once` just be mapping to `.OnMultiple` with a hard coded "1" instead?

```
// Once pass through
func (r *Events) Once(eventName string, callback func(optionalData ...interface{})) {
	// r.eventManager.Once(eventName, callback) // this does nothing new
        r.eventManager.OnMultiple(eventName, callback, 1) // 
}
```

**Reply by leaanthony** [reply]
That's what it does, just not here.
